Sources/NetworkProtection/Diagnostics/NetworkProtectionLatencyMonitor.swift

```diff
@@ -159,71 +124,29 @@ final public class NetworkProtectionLatencyMonitor {
                 let now = Date()
                 if let self,
                     (now.timeIntervalSince1970 - self.lastLatencyReported.timeIntervalSince1970 >= Self.reportThreshold) || ignoreThreshold {
-                    self.subject.send(.quality(quality))
+                    callback(.quality(quality))
                     self.lastLatencyReported = now
                 }
             }
 
-        do {
-            try await scheduleTimer()
-        } catch {
-            os_log("⚫️ Stopping latency monitor prematurely", log: log)
-            throw error
-        }
-    }
-
-    public func stop() async {
-        os_log("⚫️ Stopping latency monitor", log: log)
-        await stopScheduledTimer()
-    }
-
-    // MARK: - Timer scheduling
-
-    private func scheduleTimer() async throws {
-        await stopScheduledTimer()
-
-        await timerRunCoordinator.start()
-
-        let timer = DispatchSource.makeTimerSource(queue: timerQueue)
-        self.timer = timer
-
-        timer.schedule(deadline: .now() + Self.measurementInterval, repeating: Self.measurementInterval)
-        timer.setEventHandler { [weak self] in
-            guard let self else { return }
-
-            Task {
-                await self.measureLatency()
-            }
-        }
-
-        timer.setCancelHandler { [weak self] in
-            self?.timer = nil
+        task = Task.periodic(interval: Self.measurementInterval) { [weak self] in
+            await self?.measureLatency()
         }
-
-        timer.resume()
-    }
-
-    private func stopScheduledTimer() async {
-        await timerRunCoordinator.stop()
-
-        cancelTimerImmediately()
     }
 
-    private func cancelTimerImmediately() {
-        guard let timer else { return }
-
-        if !timer.isCancelled {
-            timer.cancel()
-        }
+    @MainActor
+    public func stop() {
+        os_log("⚫️ Stopping latency monitor", log: .networkProtectionPixel)
 
-        self.timer = nil
+        latencyCancellable = nil
+        task = nil
     }
 
     // MARK: - Latency monitor
 
     @MainActor
     public func measureLatency() async {
-        guard let serverIP = serverIP() else {
+        guard let serverIP else {
             latencySubject.send(Self.unknownLatency)
```

Comment:
If this happens the state would be badly broken, and reporting unkown latency would silently hide the issue.

Can we instead set `serverIP` in the `init` method so it's never optional?  It would clean up the code a bit and it would help us remove these lines which would make things clearer.

Reply:
`serverIP` is taken from `lastSelectedServerInfo?.ipv4` so it might not be known at initialization time. It also needs updating when server location changes. At call site, there’s a check for the ip to be present:

```
guard let ip = lastSelectedServerInfo?.ipv4 else {
     latencyMonitor.stop()
     return
}
```

Should I use an `assertionFailure` instead inside the `measureLatency()` method above to indicate a programmer error? `start(serverIP:callback:)` requires a non-nil ip anyway.

Lmk what you think.

Reply:
I'd consider adding the server IP as a non optional param to `measureLatency(to address: IPv4Address)`.  That way we don't need an optional property.